tick/optim/solver/src/svrg.cpp

```diff
@@ -1,127 +1,257 @@
     }
-
-    t += epoch_size;
+    if (delayed_updates == DelayedUpdatesMethod::Proba) {
+      solve_sparse_proba_updates();
+    }
+  } else {
+    solve_dense();
+  }
 }
 
-void SVRG::solve_sparse() {
-    // TODO: once lazy updates will be implemented in prox we will be able to
-    // do lazy updating with mu vector
-
-    // The model is sparse, so it is a ModelGeneralizedLinear and the iteration looks a
-    // little bit different
-    ulong n_features = model->get_n_features();
-    bool use_intercept = model->use_intercept();
-
-    ArrayDouble mu(iterate.size());
-    ArrayDouble fixed_w = iterate;
-    model->grad(fixed_w, mu);
-
-    ulong rand_index{0};
+void SVRG::compute_step_corrections() {
+  if (!ready_step_corrections) {
+    if ((model->is_sparse()) && delayed_updates == DelayedUpdatesMethod::Proba) {
+      ulong n_features = model->get_n_features();
+      ulong n_samples = model->get_n_samples();
+      std::shared_ptr<ModelLabelsFeatures> casted_model;
+      casted_model = std::dynamic_pointer_cast<ModelLabelsFeatures>(model);
+      ArrayULong columns_non_zeros(n_features);
+      casted_model->compute_columns_non_zeros(columns_non_zeros);
+      steps_correction = ArrayDouble(n_features);
+      for (ulong j = 0; j < n_features; ++j) {
+        steps_correction[j] = static_cast<double>(n_samples / columns_non_zeros[j]);
+      }
+      ready_step_corrections = true;
+    }
+  }
+}
 
-    if (variance_reduction == VarianceReductionMethod::Random ||
-        variance_reduction == VarianceReductionMethod::Average) {
-        next_iterate.init_to_zero();
+void SVRG::solve_dense() {
+  for (ulong t = 0; t < epoch_size; ++t) {
+    ulong i = get_next_i();
+    model->grad_i(i, iterate, grad_i);
+    model->grad_i(i, fixed_w, grad_i_fixed_w);
+    for (ulong j = 0; j < iterate.size(); ++j) {
+      iterate[j] = iterate[j] - step * (grad_i[j] - grad_i_fixed_w[j] + full_gradient[j]);
+    }
+    prox->call(iterate, step, iterate);
+    if (variance_reduction == VarianceReductionMethod::Random && t == rand_index) {
+      next_iterate = iterate;
     }
+    if (variance_reduction == VarianceReductionMethod::Average) {
+      next_iterate.mult_incr(iterate, 1.0 / epoch_size);
+    }
+  }
+  if (variance_reduction == VarianceReductionMethod::Last) {
+    next_iterate = iterate;
+  }
+  t += epoch_size;
+}
 
-    if (variance_reduction == VarianceReductionMethod::Random) {
-        rand_index = rand_unif(epoch_size);
+void SVRG::solve_sparse_proba_updates() {
+  // Data is sparse, and we use the probabilistic update strategy
+  // This means that the model is a child of ModelGeneralizedLinear.
+  // The strategy used here uses non-delayed updates, with corrected
+  // step-sizes using a probabilistic approximation and the
+  // penalization trick: with such a model and prox (if prox is separable),
+  // we can work only inside the current support (non-zero values) of
+  // the sampled vector of features
```

Comment:
Maybe can we add a link to ASAGA paper or any other paper that introduced this update type?

Reply:
This is part of the task about solvers documentation (that I started already)

Reply:
Alright!